static/js/components/Loading.js

```diff
@@ -1,27 +1,48 @@
 // @flow
 import React from "react"
+import R from "ramda"
 
-import { anyProcessing, anyError, allLoaded } from "../util/rest"
-
-import type { RestState } from "../flow/restTypes"
-
-export default class Loading extends React.Component {
-  props: {
-    restStates: Array<RestState<*>>,
-    renderContents: Function
-  }
+type LoadingProps = {
+  loaded: boolean,
+  errored: boolean
+}
 
-  render() {
-    const { restStates, renderContents } = this.props
+const withLoading = R.curry((
+  shouldLoadData: (Object, Object) => boolean,
+  loadData: Object => void,
+  LoadedComponent: Class<React.Component<*, *, *>>
+) => {
+  return class extends React.Component {
+    props: LoadingProps
 
-    if (anyProcessing(restStates) || !allLoaded(restStates)) {
-      return <div>Loading</div>
+    componentWillMount() {
+      loadData(this.props)
     }
 
-    if (anyError(restStates)) {
-      return <div>Error loading page</div>
+    componentDidUpdate(prevProps: Object) {
+      if (shouldLoadData(prevProps, this.props)) {
+        loadData(this.props)
+      }
     }
 
-    return renderContents()
+    render() {
+      const { loaded, errored, ...props } = this.props
+
+      if (errored) {
+        return <div className="errored">Error loading page</div>
+      }
+
+      if (!loaded) {
+        return <div className="loading">Loading</div>
+      }
+
+      return (
+        <div className="loaded">
+          <LoadedComponent {...props} />
+        </div>
+      )
+    }
   }
-}
+})
+
+export default withLoading
```

Comment:
if we define the HOC like this, I believe it's functionally equivalent and we could move those lifecycle methods back into the original containers:

```javascript
const withLoading = (
  LoadedComponent: Class<React.Component<*, *, *>>,
) => {
  return class extends LoadedComponent {
    props: LoadingProps

    render() {
      const { loaded, errored, ...props } = this.props

      if (errored) {
        return <div className="errored">Error loading page</div>
      }

      if (!loaded) {
        return <div className="loading">Loading</div>
      }

      return (
        <div className="loaded">
          { super.render(props) }
        </div>
      )
    }
  }
}
```

@aliceriot, any thoughts on that? Might be some drawbacks in terms of using inheritance, but I really strongly prefer this. The lifecycle methods exist in a place that you expect them to exist, and the original container still functions on its own with or without the loading HOC (not so with the current approach)